### PR TITLE
added hw-4 implementation

### PR DESCRIPTION
**Pull request by oroboro-cpu**

### CODE REVIEW

File: src/main/java/project/service/CarServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public boolean delete(Long id) {
+        return carDao.delete(id);
+    }
+
+    @Override
+    public void addDriverToCar(Driver driver, Car car) {
+        List<Driver> drivers = new ArrayList<>();
+        drivers.add(driver);
+        car.setDrivers(drivers);
+        carDao.update(car);
```
Comment (yurwing) [inline src/main/java/project/service/CarServiceImpl.java]: it seems you are rewriting the data here, respectively, only one driver can be attached to each car

**Review by yurwing — CHANGES_REQUESTED** [review]

**Review by yurwing — APPROVED** [review]

File: .travis.yml [hunk 2]
```diff
@@ -5,4 +5,4 @@ script:
   - mvn clean install
 cache:
   directories:
-    - $HOME/.m2
+    - $HOME/.m2
```
Comment (Sofasmile) [inline .travis.yml]: let's add empty line at the end of the file

File: src/main/java/project/dao/CarDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public boolean delete(Long id) {
+        return Storage.cars.removeIf(c -> c.getId().equals(id));
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        return getAll().stream()
+                .filter(c -> c.getDrivers().stream()
+                        .map(Driver::getId)
```
Comment (Sofasmile) [inline src/main/java/project/dao/CarDaoImpl.java]: try to use method anyMatch() here

File: src/main/java/project/models/Car.java [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
 public class Car {
     private Long id;
     private String name;
     private Manufacturer manufacturer;
     private List<Driver> drivers;
 
+    public Car(String name, Manufacturer manufacturer, List<Driver> drivers) {
+        this.name = name;
+        this.manufacturer = manufacturer;
+        this.drivers = drivers;
```
Comment (Sofasmile) [inline src/main/java/project/models/Car.java]:
```suggestion
    public Car(String name, Manufacturer manufacturer) {
        this.name = name;
        this.manufacturer = manufacturer;
        this.drivers = new ArrayList();
```

File: src/main/java/project/dao/CarDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+@Dao
+public class CarDaoImpl implements CarDao {
+    @Override
+    public Car create(Car car) {
+        Storage.addCar(car);
+        return car;
+    }
+
+    @Override
+    public Car get(Long id) {
```
Comment (Sofasmile) [inline src/main/java/project/dao/CarDaoImpl.java]:
```suggestion
    public Optional<Car> get(Long id) {
```

File: src/main/java/project/dao/DriverDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+@Dao
+public class DriverDaoImpl implements DriverDao {
+    @Override
+    public Driver create(Driver driver) {
+        Storage.addDriver(driver);
+        return driver;
+    }
+
+    @Override
+    public Driver get(Long id) {
```
Comment (Sofasmile) [inline src/main/java/project/dao/DriverDaoImpl.java]: the same

File: src/main/java/project/service/CarServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        List<Driver> driverList = new ArrayList<>(car.getDrivers());
+        Driver driverToRemove = car.getDrivers().stream()
+                .filter(d -> d.equals(driver))
+                .findFirst()
+                .orElseThrow();
+        driverList.remove(driverToRemove);
```
Comment (Sofasmile) [inline src/main/java/project/service/CarServiceImpl.java]:
```suggestion
        driverList.remove(driver);
```

**Review by Sofasmile — CHANGES_REQUESTED** [review]

File: src/main/java/project/service/CarService.java [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
 import java.util.List;
+import java.util.Optional;
 import project.models.Car;
 import project.models.Driver;
 
 public interface CarService {
     Car create(Car car);
 
-    Car get(Long id);
+    Optional<Car> get(Long id);
```
Comment (Sofasmile) [inline src/main/java/project/service/CarService.java]: On the service layer we shouldn't return Optional, only on the dao layer

**Review by Sofasmile — CHANGES_REQUESTED** [review]

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**